README.md

```diff
@@ -17,3 +17,380 @@ Implement a prototype job worker service that provides an API to run arbitrary L
+
+* when the job is done it will generate a log called END-\<endtimestamp\>.log, and the process table will be updated
+
+```
+type StartRequest {
+    //path of a program or executable
+    string command
+    //arguments to invoke program or executable with
+    []string args 
+    // environment variables for the execution
+    //OPTIONAL variable
+    []string env
+    // current working directory of the execution
+    //OPTIONAL variable, default will be the working directory of server 
+    string dir 
+}
+
+type StartResponse {
+    //process Id of the running process executed by the command
+    //will be 0 if process failed to execute
+    int pid 
+    //univeral unique identifier that tags each unique request made to server
+    string uuid
+    //starting time of start request
+    string startTimeStamp
+
+}
+
+func ExecuteStart(StartRequest) returns(StartResponse)
+```
+
+#### Stop
+
+* The User should be able to stop the request based on the pid
+
+* When stopped the pid should be killed 
+
+* A response will be sent to client indicating process have been stopped along with the contents of the log
+
+* job should be marked as completed with the exit code in the dataStore
+
+```
+type StopRequest {
+    int pid
+}
+
+type StopResponse {
+    []byte stdout
+    []byte stderr
+    bool isKilled
+}
+
+func ExecuteStop(StopRequest) returns(StopResponse)
+
+```
+
+#### Query 
+
+
+* There will be only two types of Query
+
+    * QueryOneProcess:
+
+        * return the logs of a job using a valid pid (if it was started) or uuid, along with ProcessInfo
+
+
+    * QueryRunningProcesses:
+
+        * get a list of job's executed and processInfo for the jobs
+
+```
+type ProcessInfo {
+    int pid 
+    string startTimeStamp 
+    string endTimeStamp
+    string processName 
+    string uuid
+    bool isRunning 
+    int exitCode
+}
+
+type QueryOneProcessRequest {
+    int pid 
+    string uuid
+}
+
+type QueryOneProcessResponse {
+    processInfo procInfo
+    []byte stdout 
+    []byte stderr 
+}
+
+type QueryRunningProcessesRequest {
+    //will be empty since the server just needs to verify its a QueryRunningProcessesRequest
+}
+    
+type QueryRunningProcessesResponse {
+     ProcessInfo[] processTable 
+}
+
+func ExecuteQueryOneProcess(QueryOneProcessRequest) returns(QueryOneProcessResponse)
+func ExecuteQueryRunningProcesses(QueryRunningProcessesRequest) returns(QueryRunningProcessesResponse)
+
+```
+
+
+### Implementation Overview 
+
+* Using GRPC so the above methods and types with be generated via protocol buffers (will be using libprotoc 3.11.4)
+
+* the rpc APIs' will be all unary to keep the client - server communication simple
+
+####  DataStore
+
+* we can use Map in Go to implement a set of structs to store process info and use sync.mutex to handle concurrent transactions, the key to the map will be \<uuid\>-\<startTimeStamp\>
```

Comment:
Why use `startTimeStamp` in the map key?

Reply:
will change that to just using uuid